### PR TITLE
Feature/firebase login

### PR DESCRIPTION
**Pull request by yanvow**
- Implement email/password authentication using firebase
- added a log out button in the menu drawer

### CODE REVIEW

**Review by schuetzcarl — CHANGES_REQUESTED** [review]
I inspected the whole Login implementation and globally it is a great work. Thanks for your application and making such a modular code, it will be really useful later.
Even if it deserve a bit more documentation (for example, the login() method should be more documented, as it looks like a nice pattern with three call level) and some lines should be grouped and their common action should be described in comment in order to be more readable, I don't want to bother you with it now as having a better test coverage is the priority before any merge.

One thing that I would advise in the architecture, is to create a `LoginDataSource` Interface with an abstract `login()` method. And rename your `LoginDataSource` class into `FirebaseLoginDataSource` which would implement the latter.
With that we can easily create some other `LoginDataSource` (Maybe a mock one for testing, even if mockito already does the job), and chose which concrete instance we want in the `LoginViewModel` class.

One last thing, is that the class organization would need a small refactor to bind with the `UserViewModel,` but it will be more clear, and better if I show you my proposition in a concrete example to resolve the issue #66 in a next PR with you as reviewer.

Hence, after having a better test coverage, I will agree to merge your work.

**Review by schuetzcarl — CHANGES_REQUESTED** [review]
Congrats for fixing your testing issue. As mentioned earlier, your login implementation is a great work, with a smart and appropriate use of some Design Pattern.

However, I think that some .kt files should be removed, and Classes Or Kotlin objects written in moved as inner class/object from the class they are related to (Factory example). It really add more readability in the project tree files.
As I will need to change a bit some of these classes to properly bind your work with the NavFragment and Menu (see issue : #66) to avoid useless work, don't change it now. I will change that during the latter issue and ask yourself as reviewer.

Note that I found a bug during my manual testing part of the review :
When I write a non valid user login, click confirm and retry with the valid login, the login doesn't work. Or when I try several times to enter a wrong entry, and finally want to access with the correct, it fails.
I suggest you to write a regression test and then resolve that scenario in a new opened issue. (See with the product owner if it need to be resolved in that sprint or not)


Hence, to sum up, I agree to merge your work immediately after you resolved my few in-files review comments.

File: app/src/main/java/com/github/factotum_sdp/factotum/data/LoginRepository.kt
[the remark this answers is not in the document]
Comment (yanvow) [reply app/src/main/java/com/github/factotum_sdp/factotum/data/LoginRepository.kt:1]: `private set` makes the user variable's setter private.
Isn't it sufficient?

[the remark this answers is not in the document]
Comment (schuetzcarl) [reply app/src/main/java/com/github/factotum_sdp/factotum/data/LoginRepository.kt:1]: Yes it is, sorry I missed it.

[the remark this answers is not in the document]
Comment (schuetzcarl) [reply app/src/main/java/com/github/factotum_sdp/factotum/data/LoginRepository.kt:1]: But the point is finally more about your non-android tests, they are for most of them seems not very relevant, even for coverage. 
As now you managed more front-end tests, can you filter the useless back-end tests and turn private all the login class variables read only in the tests ?

**Comment by yanvow** [comment]
Concerning the bug you had, it's not on my side but rather on Firebase's. There seems to be a cool down after incorrect login information. Basically, you just need to wait before retrying with the correct information. 
I could implement a cool down for the button but I would rather do it in a future task. Also taking into account the priority of have an up and running login even with minor flaws.

**Comment by schuetzcarl** [comment]
Nice to already have locate from where the problem come from.

Yes that's what I suggested to solve this in an other issue and not necessarily now.

File: app/src/main/java/com/github/factotum_sdp/factotum/data/LoginRepository.kt
[the remark this answers is not in the document]
Comment (yanvow) [reply app/src/main/java/com/github/factotum_sdp/factotum/data/LoginRepository.kt:1]: The login class variables are already private or with private setters

[the remark this answers is not in the document]
Comment (schuetzcarl) [reply app/src/main/java/com/github/factotum_sdp/factotum/data/LoginRepository.kt:1]: I mean the "login" related classes with for example `LoginRepository` included, and his 3 atributes `dataSource`, `user` and `isLoggedIn` which have to be private.

**Review by schuetzcarl — APPROVED** [review]
Looks good to me for merging, thank you for the changes made.

By the way, try to avoid such a fine granularity in your commits. There is already enough commit waste due to the finicky CI

**Pull request merged**